text/0000-glimmer-components.md

```diff
@@ -0,0 +1,1019 @@
+element. Given this component template:
+
+```hbs
+Hello, world!
+```
+
+The output by _default_ would be something like:
+
+```html
+<!-- OUTPUT -->
+<div id="ember-1234" class="ember-view">
+  Hello, world!
+</div>
+```
+
+But we can't know that for sure unless we look at the component definition. If
+we do, we might see that the outer wrapping element is actually a `section`, and
+it has a `.hello-world` class:
+
+```js
+export default Component.extend({
+  tagName: 'section',
+  classNames: ['hello-world']
+});
+```
+```html
+<!-- OUTPUT -->
+<section id="ember-1234" class="hello-world ember-view">
+  Hello, world!
+</section>
+```
+
+This behavior means that the template for a component is missing crucial
+information and context. Even for the simplest component, users must check the
+class definition to know with certainty what the full template of the component
+is. And unlike bindings, there is no hint to the user that there may be
+something dynamic that they should check on - without advanced knowledge of
+Ember's APIs, there is no way of knowing about this behavior.
+
+By contrast, Glimmer components have no wrapping outer element - What you see in
+the template is what you get in the output. There is no way to define class
+names, class name bindings, attribute bindings, or any other DOM element values
+_from the component class_. The template is the single source of truth for the
+output of a component, and any dynamic values are explicitly stated in it.
+
+```hbs
+<!-- template.hbs -->
+<section class="hello-world">
+  Hello, world!
+</section>
+```
+
+```html
+<!-- OUTPUT -->
+<section class="hello-world">
+  Hello, world!
+</section>
+```
+
+We can immediately see that this is a static component with no bindings, no
+dynamic values, and no meaningful state. Even if there was a component
+definition, we know that it is not in any way affecting the output of this
+template. Special element ids and classes are also not present, making the
+output appear less magical.
+
+This small change makes a huge difference in the large: users can spend less
+time switching back and forth between reading template and class code, and can
+get a better idea of the structure of an app from its declarative templates;
+thinner classes are incentivized, guiding users toward putting more of their app
+in templates and away from programmatic/imperative style design; and the gap
+between template-only components - which are analogous to React and other
+frameworks' functional components - and components with a backing class is
+reduced, making them a more viable pattern.
+
+### Namespaced Arguments
+
+In classic components, arguments are set as properties directly on the class
+instance. This means that class methods and properties can be completely
+overwritten by incoming arguments, which can have surprising and problematic
+side effects. For example, let's say we have a component that has a
```

Comment:
> This means that class methods and properties can be completely
overwritten by incoming arguments, which can have surprising and problematic
side effects.

This isn't always problematic. The same phenomenon can be helpful for default values

```hbs
  {{cart-item price=5.25 }}
  {{cart-item price=1.99 qty=3 }}
```
```js
export default Ember.Component.extend({
   qty: 1,
   total: computed('price', 'qty', function() {
     return this.price * this.qty;
   })
});
```

Reply:
This is true, and something we should explicitly address in drawbacks.